specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts

```diff
@@ -46,6 +47,48 @@ type MappedBusinessRuleDefs = {
 };
 
 export const businessRuleDefs: MappedBusinessRuleDefs = {
+  Address: {
+    customInit: (address) => {
+      if (address.isNew()) {
+        const setPrimary = (): void => {
+          address.set('isPrimary', true);
+          if (address.collection !== undefined) {
+            address.collection.models.forEach(
```

Comment:
A `resource.collection` is any instance of a BackBone Collection. 

In Specify, these are generally subclasses of Base, which is a subclass of BackBone.Collection https://github.com/specify/specify7/blob/94fc650f1121e021dfe32565d4c2039c38da1ff6/specifyweb/frontend/js_src/lib/components/DataModel/collectionApi.js#L7-L12

The type is defined here: 
https://github.com/specify/specify7/blob/94fc650f1121e021dfe32565d4c2039c38da1ff6/specifyweb/frontend/js_src/lib/components/DataModel/specifyTable.ts#L85-L113

`resource.collection` can also be undefined (not exist on the Resource) if it is not present in any Collection. I have not encountered any instance where `resource.collection` has been null, but there were places in the code where the check has been made. 

In the application this is most often a Dependent Collection, although all resources on the 'Many" side of a `to-many` relationship should have a defined Collection.  

![image](https://github.com/specify/specify7/assets/64045831/c0ac1355-ecc3-46ba-b38b-d16b2df112d6)

And I don't think it is used on the frontend anywhere, but `resource.collection` is also defined on resources fetched from a `LazyCollection` or a `ToOneCollection`. 

For instance, when fetching the Picklist Items for picklists of type 'Entire Table', every fetched resource will be in the same Collection. 

https://github.com/specify/specify7/blob/94fc650f1121e021dfe32565d4c2039c38da1ff6/specifyweb/frontend/js_src/lib/components/FormCells/PickListEditor.tsx#L49-L68

`resource.collection.models` specifically is an array of `SpecifyResource` objects which corresponds to every resource in the collection.

Reply:
thanks for the detailed explanation!
regarding https://github.com/specify/specify7/issues/4286, I am curious if it would be veasable in the new resource API to replace all kinds of Collections we have right now (especially the StaticCollection) with just a plain array, or if there are some important things we loose in the process

for things like pagination, we can provide higher level utilities that do that (in fact this logic is already implemented in the attachments page, query results, record set - we should deduplicate it if possible)
then of course, collection has indication of what it's parent resource and the relating relationship is - in cases where that data is needed, can it be passed as a separate argument?
and perhaps the most important thing collection does is enforce integrity (if subresource is changed, parent resource is notified of the change, and if resource is saved, appropriate resources are notified about the change and react accordingly) - I do not know if this can be replaced easily but would depend on the architecture we settle with

anyway - not saying collections are not useful, just want to explore if it's possible to do without them and if that would simplify or complicate things

Reply:
In most cases, I agree it would be easier to replace Backbone Collections with arrays. 

The most useful attributes on a Collection are likely: 
- field
  - If not undefined, stores the associated Relationship from a Parent resource to the Collection from the side of the Collection Models (for instance the `collectionObject` relationship on Determinations).
- related
  - stores a reference to the Parent (or otherwise related) resource. Instead of awaiting a `determination.rget(collectionObject)`, if the determination is in a Collection then `determination.collection.related` would synchronously return the CollectionObject as the CollectionObject is already stored in the Collection.related.
- models
  - An array containing every resource in the Collection 
- length (should be same as `addmodels.length`)
- _events
  - While this should not be directly used, it holds a reference to every Backbone event the Collection is listening to.

To my knowledge, the places were Collection events are relevant are in:

resourceApi.js
https://github.com/specify/specify7/blob/7c6b8e5b736a369ebfad5b45b4cb0d1ec29c9ba8/specifyweb/frontend/js_src/lib/components/DataModel/resourceApi.js#L50-L72

collectionApi.js
https://github.com/specify/specify7/blob/7c6b8e5b736a369ebfad5b45b4cb0d1ec29c9ba8/specifyweb/frontend/js_src/lib/components/DataModel/collectionApi.js#L44-L75

businessRules.ts
https://github.com/specify/specify7/blob/7c6b8e5b736a369ebfad5b45b4cb0d1ec29c9ba8/specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts#L52-L53

And a few places in [businessRuleDefs.ts](https://github.com/specify/specify7/blob/production/specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts)

As you have stated, accommodating these functionalities will depend on what we eventually settle with. The uses of `collection.related` would most likely be solved if we stored/cached fetched Resources. 
There are few usages of Collection events, but we would still need some way to replicate or otherwise modify the current behavior. Whether this can be accomplished easily depends on the design we eventually settle with. At the very least, most of the "bloat" of Collections can be removed in favor of a simpler approach.